python-frontend/src/main/java/org/sonar/python/api/IPythonGrammarBuilder.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * SonarQube Python Plugin
+ * Copyright (C) 2011-2023 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.python.api;
+
+import org.sonar.sslr.grammar.LexerfulGrammarBuilder;
+
+import static com.sonar.sslr.api.GenericTokenType.EOF;
+import static org.sonar.python.api.IPythonGrammar.LINE_MAGIC_STATEMENT;
+import static org.sonar.python.api.IPythonGrammar.LINE_MAGIC;
+import static org.sonar.python.api.PythonGrammar.ASSERT_STMT;
+import static org.sonar.python.api.PythonGrammar.ANNOTATED_RHS;
+import static org.sonar.python.api.PythonGrammar.BREAK_STMT;
+import static org.sonar.python.api.PythonGrammar.CONTINUE_STMT;
+import static org.sonar.python.api.PythonGrammar.DEL_STMT;
+import static org.sonar.python.api.PythonGrammar.EXEC_STMT;
+import static org.sonar.python.api.PythonGrammar.EXPRESSION_STMT;
+import static org.sonar.python.api.PythonGrammar.GLOBAL_STMT;
+import static org.sonar.python.api.PythonGrammar.IMPORT_STMT;
+import static org.sonar.python.api.PythonGrammar.NAME;
+import static org.sonar.python.api.PythonGrammar.NONLOCAL_STMT;
+import static org.sonar.python.api.PythonGrammar.PASS_STMT;
+import static org.sonar.python.api.PythonGrammar.PRINT_STMT;
+import static org.sonar.python.api.PythonGrammar.RAISE_STMT;
+import static org.sonar.python.api.PythonGrammar.RETURN_STMT;
+import static org.sonar.python.api.PythonGrammar.SIMPLE_STMT;
+import static org.sonar.python.api.PythonGrammar.TESTLIST_STAR_EXPR;
+import static org.sonar.python.api.PythonGrammar.YIELD_EXPR;
+import static org.sonar.python.api.PythonGrammar.YIELD_STMT;
+import static org.sonar.python.api.PythonTokenType.NEWLINE;
+
+public class IPythonGrammarBuilder extends PythonGrammarBuilder {
+
+  @Override
+  protected void setupRules(LexerfulGrammarBuilder b) {
+    lineMagic(b);
+    super.setupRules(b);
```

Comment:
In hindsight, I have some doubt about this, we are registering the magic line rule with the highest priority, and maybe it would be better to completely rewrite the `setupRules` method for more fine tuned priority.
It happens to work this way though, so I'm not sure we need to change anything. @andreaguarino  wdyt?